test/hash/src/test_hash.d

```diff
@@ -232,6 +233,19 @@ void issue20034()
     assert(hashOf(E.a, 1));
 }
 
+/// [REG 2.084] hashOf will fail to compile for some structs/unions that recursively contain shared enums
+void issue21642() @safe nothrow pure
+{
+    enum C : char { _ = 1, }
+    union U { C c; void[0] _; }
+    shared union V { U u; }
+    cast(void) hashOf(V.init);
+    // Also test the underlying reason the above was failing.
+    import core.internal.convert : toUbyte;
+    shared C c;
+    cast(void) toUbyte(c);
```

Comment:
Instead of casting to void here and above, it's better to check if the return value is the expected one.

Reply:
Added a check here. Above there is no particular value the hash code is supposed to be.

Reply:
My motivation is the question: why would an optimizing compiler not remove a call to a pure function, if it's result is not used?

Technically such optimization happens after semantic analysis (where the bug manifests), in the backend, but in theory the frontend could also discard such function calls, so it's better to make the code more robust.

Reply:
Is there a change you would suggest that wouldn't make the test brittle?

Reply:
@PetarKirov Pure functions that may throw exceptions should not be optimized away. This pattern is prevalent in druntime: a function that does some checks and throws if the input is not the expected one.